File: backend/engine/terms.py
"""
Academic term ordering.

The planner generates Fall/Spring plans, but students may add optional Summer or
Winter terms (as manual relief valves — drag a class there to lighten a Fall/Spring
load). For prerequisite/standing checks to stay correct, the plan's semesters must
be processed in true chronological order. This module is the single source of that
order, used by the API's revalidation step (and mirrored in the frontend).

SDSU calendar order WITHIN a calendar year:
    Winter (January intersession) → Spring → Summer → Fall

So a course in Fall 2026 precedes Winter 2027, which precedes Spring 2027, etc.
"""

from typing import Any

# Rank of each term within its calendar year (lower = earlier).
TERM_RANK: dict[str, int] = {"Winter": 0, "Spring": 1, "Summer": 2, "Fall": 3}

# Terms a student may add on top of the generated Fall/Spring plan.
OPTIONAL_TERMS: tuple[str, ...] = ("Winter", "Summer")


def chronological_key(semester: str, year: int) -> tuple[int, int]:
    """Sort key placing (semester, year) in true calendar order.

    Unknown term names sort last within their year (rank 9) so bad data never
    silently reorders real terms ahead of themselves.
    """
    return (int(year), TERM_RANK.get((semester or "").strip().title(), 9))


def sort_semesters(semesters: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Return the plan's semesters in chronological order (stable)."""
    return sorted(
        semesters,
        key=lambda s: chronological_key(s.get("semester", ""), s.get("year", 0)),
    )

File: frontend/src/utils/terms.js
// Academic term ordering — mirror of backend/engine/terms.py.
// SDSU calendar order within a year: Winter → Spring → Summer → Fall.

export const TERM_RANK = { Winter: 0, Spring: 1, Summer: 2, Fall: 3 };
export const OPTIONAL_TERMS = ["Winter", "Summer"];

export function chronoKey(semester, year) {
  return Number(year) * 10 + (TERM_RANK[semester] ?? 9);
}

export function sortSemesters(semesters) {
  return [...semesters].sort(
    (a, b) => chronoKey(a.semester, a.year) - chronoKey(b.semester, b.year)
  );
}

// Optional (Summer/Winter) terms that could be inserted into the current plan but
// aren't there yet — one suggestion per gap, in chronological position.
// Winter Y sits between Fall (Y-1) and Spring Y; Summer Y between Spring Y and Fall Y.
export function suggestedTerms(semesters) {
  const present = new Set(semesters.map((s) => `${s.semester} ${s.year}`));
  const out = [];
  for (const s of semesters) {
    if (s.semester === "Spring") {
      const k = `Summer ${s.year}`;
      if (!present.has(k)) out.push({ semester: "Summer", year: s.year });
    }
    if (s.semester === "Fall") {
      const k = `Winter ${s.year + 1}`;
      if (!present.has(k)) out.push({ semester: "Winter", year: s.year + 1 });
    }
  }
  // De-dup and order chronologically.
  const seen = new Set();
  return sortSemesters(
    out.filter((t) => {
      const k = `${t.semester} ${t.year}`;
      if (seen.has(k)) return false;
      seen.add(k);
      return true;
    })
  );
}
